File: backend-old/routes/plugins.js
const express = require('express');
const router = express.Router();
const pluginService = require('../services/pluginService');
const minecraftService = require('../services/minecraftService');

const { auth, checkPermission } = require('../middleware');

// Search Plugins
router.get('/search', auth, async (req, res) => {
    try {
        const { query } = req.query;
        if (!query) {
            return res.status(400).json({ error: 'Query parameter is required' });
        }
        const results = await pluginService.search(query);
        res.json(results);
    } catch (err) {
        console.error('Search error:', err);
        res.status(500).json({ error: 'Failed to search plugins' });
    }
});

// Install Plugin
router.post('/install', auth, checkPermission('plugins.manage'), async (req, res) => {
    try {
        const { projectId, source = 'Modrinth' } = req.body;

        // Get current server config automatically
        const config = minecraftService.config;
        const version = config.version; // e.g., '1.20.4'
        let type = config.type || 'vanilla';

        // Map server type to compatible Modrinth loaders
        const loaderMap = {
            'paper': ['paper', 'purpur', 'spigot', 'bukkit'],
            'purpur': ['purpur', 'paper', 'spigot', 'bukkit'],
            'spigot': ['spigot', 'bukkit', 'paper'],
            'bukkit': ['bukkit', 'spigot'],
            'vanilla': ['paper', 'purpur', 'spigot', 'bukkit'] // Fallback
        };

        const loaders = loaderMap[type] || ['paper', 'purpur', 'spigot', 'bukkit'];

        if (!projectId) {
            return res.status(400).json({ error: 'Project ID is required' });
        }

        console.log(`Installing plugin ${projectId} from ${source} for ${version} (loaders: ${loaders.join(', ')})`);

        const result = await pluginService.install(source, projectId, version, loaders);
        res.json({ success: true, ...result });

    } catch (err) {
        console.error('Install error:', err);
        res.status(500).json({ error: err.message });
    }
});

module.exports = router;

File: backend-rust/src/routes/plugins.rs
use std::sync::Arc;

use axum::{
    extract::State,
    routing::{get, post},
    Json, Router,
};
use serde::Deserialize;

use crate::{
    error::Result,
    middleware::{AuthUser, PermissionGuard},
    services::{InstallResult, PluginInfo},
    state::AppState,
};

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/search", get(search_plugins))
        .route("/install", post(install_plugin))
}

#[derive(Debug, Deserialize)]
struct SearchQuery {
    #[serde(default)]
    query: String, // Changed from q to query directly
    #[serde(default = "default_limit")]
    limit: usize,
}

fn default_limit() -> usize {
    10
}

async fn search_plugins(
    State(state): State<Arc<AppState>>,
    auth_user: AuthUser,
    axum::extract::Query(params): axum::extract::Query<SearchQuery>,
) -> Result<Json<Vec<PluginInfo>>> {
    PermissionGuard::check(&state, &auth_user, "plugins.view").await?;
    
    // Use params.query instead of params.q
    let results = state.plugin.search(&params.query, params.limit).await?;
    Ok(Json(results))
}

#[derive(Debug, Deserialize)]
struct InstallRequest {
    source: String,
    #[serde(alias = "projectId")]
    id: String,
    #[serde(default)]
    version: Option<String>,
    #[serde(default)]
    loaders: Vec<String>,
}

async fn install_plugin(
    State(state): State<Arc<AppState>>,
    auth_user: AuthUser,
    Json(payload): Json<InstallRequest>,
) -> Result<Json<InstallResult>> {
    PermissionGuard::check(&state, &auth_user, "plugins.install").await?;
    
    let loaders = if payload.loaders.is_empty() {
        vec!["paper".to_string(), "bukkit".to_string(), "spigot".to_string()]
    } else {
        payload.loaders
    };
    
    let version_str = payload.version.unwrap_or_default();
    
    let result = state
        .plugin
        .install(&payload.source, &payload.id, &version_str, &loaders)
        .await?;
    
    Ok(Json(result))
}
